pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -210,19 +231,134 @@ pub(super) fn spawn_storage_sync_thread<
     remote_storage: S,
-    concurrent_sync_limit: &Semaphore,
+    max_concurrent_sync: NonZeroUsize,
+    max_sync_errors: NonZeroU32,
 ) -> anyhow::Result<()> {
     let runtime = tokio::runtime::Builder::new_current_thread()
         .enable_all()
         .build()?;
-    let mut remote_timelines = runtime
+    let remote_timelines = runtime
         .block_on(fetch_existing_uploads(&remote_storage))
         .context("Failed to determine previously uploaded timelines")?;
 
-    let urgent_downloads = latest_timelines(&remote_timelines)
+    schedule_first_tasks(config, &remote_timelines);
+
+    let remote_timelines = Arc::new(Mutex::new(remote_timelines));
+    let remote_storage = Arc::new(remote_storage);
+
+    while !crate::tenant_mgr::shutdown_requested() {
+        let next_tasks = get_next_tasks(max_concurrent_sync);
+        let remaining_queue_length = SYNC_QUEUE.len();
+
+        if next_tasks.is_empty() {
+            trace!("No storage sync tasks found");
+            thread::sleep(Duration::from_secs(1));
+            continue;
+        } else {
+            debug!(
+                "Processing {} tasks, more tasks left to process: {}",
+                next_tasks.len(),
+                remaining_queue_length
+            );
+            REMAINING_SYNC_ITEMS.set(remaining_queue_length as i64);
+
+            let mut task_batch = next_tasks
+                .into_iter()
+                .map(|task| {
+                    tokio::spawn(process_task(
+                        config,
+                        Arc::clone(&remote_storage),
+                        task,
+                        max_sync_errors,
+                        Arc::clone(&remote_timelines),
+                    ))
+                })
+                .collect::<FuturesUnordered<_>>();
+            runtime.block_on(async move {
+                while let Some(task) = task_batch.next().await {
+                    if let Err(e) = task {
+                        error!("Failed to schedule storage sync task: {:#}", e);
+                    }
+                }
+            });
+        }
+    }
+    debug!("Queue based remote storage sync thread shut down");
+    Ok(())
+}
+
+async fn process_task<
+    P: std::fmt::Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+>(
+    config: &'static PageServerConf,
+    remote_storage: Arc<S>,
+    task: SyncTask,
+    max_sync_errors: NonZeroU32,
+    remote_timelines: Arc<Mutex<HashMap<(ZTenantId, ZTimelineId), RemoteTimeline>>>,
+) {
+    if task.retries > max_sync_errors.get() {
+        error!(
+            "Evicting task {:?} that failed {} times, exceeding the error theshold",
+            task.kind, task.retries
+        );
+        let tenant_id = task.tenant_id();
+        let timeline_id = task.timeline_id();
+        if let Err(e) =
+            get_repository_for_tenant(tenant_id).and_then(|repo| repo.unload_timeline(timeline_id))
```

Comment:
Another question is how adequate it is to unload a timeline, if we fail to download its updates or upload some of its checkpoints.

Both failures may occur unpredictably and always occur in the background, so a very sudden timeline disappear might happen due to this.

Reply:
Will unload_timeline be blocking if it waits for it to unload

Reply:
I don't think it will impact much the other tasks, since we wrap every `process_task` invocation into `tokio::spawn` that has to mitigate such effects, but sure the batch processing per se may be affected or even get blocked.

We could do that in a separate thread explicitly, not sure how feasible to do.

Reply:
I'm worried that unload_timeline will `join()` on other threads which may take a while. This could block the storage sync executor right?

Reply:
Yes, partially, if we use a multithreaded tokio runtime, but that won't much of a solace.

I think, we could batch such tasks and them register them all at once, which can be done in the end, this blocking in the end of the loop. Not superb, but better than spawning a thread on the loop?

Reply:
Could we also use `spawn_blocking`

Reply:
We could, but the batch register seemed simple to implement, that's what I've ended up with 🙂 
While we can use `spawn_blocking` on it still, makes not much sense IMO: the batch register is in the end of the loop iteration and we'll have to wait at the `spawn_blocking` join handle at some point: making things more async than they are might be cumbersome.

Reply:
I think my concern is that we're calling into the sync world of LayeredRepo from async functions. unload_timeline blocks until the timeline is removed from memory and all associated tasks are shut down, which seems like it could take a while. Is my understanding of this wrong?

Reply:
No, you're absolutely right here, it's all done sync and could take a while, despite some attempts to do things in batch now.

There's no `async` world around the `register_timeline_downloads` though, that's why I think it's "fine": it's done outside of `runtime.block_on` calls.
I still agree that's not ideal, since might still take a while, even though being outside async machinery, not blocking it directly.

I don't have ready plans to tackle such "a while" things besides profiling and rewriting the registration code?
We could move off the registration into another thread/thread-wrapped-in-future and loop on, as I've mentioned.
But that would require managing two queues now, not sure if that's a nice or easy way to go with.
No good ideas on the topic from me so far.

Reply:
I'm thinking about the call to `unload_timeline` done inside the async fn `process_task` which is executed by the tokio::spawn call in `loop_step`.

Reply:
Now get it, thanks, very good catch.

Yes, that's an issue, I've moved it away, similarly to what I've done to downloads.
It gets even better now, since `tenant_mgr` has a generic method to register any future sync operations, if needed.